### DISCUSSION TITLE
HTTP GET requests to monitoring on failures of the input and output

### DISCUSSION THREAD

**Discussion by KlaasT** [savonet/liquidsoap#2960] [User Support]
Dear all,

I've been searching through the docs and issues as I am trying to add some monitoring to my liquidsoap instances. 

What I would love to do is fire a HTTP GET request to a monitoring API to notify if for example the input/source fails for some reason or when the output encoder has connection issues or other problems.

Is this actually possible?

Thanks very much. Every piece of help is highly appreciated.

**Comment by toots** [comment]
Hi @KlaasT ! 

Yes, this is absolutely possible!

The first place to look is our `prometheus` support: https://www.liquidsoap.info/doc-2.1.4/reference.html#prometheus.counter

Unfortunately, this support is under-document by, essentially, each operator gives you the ability to register a specific prometheus metric. The returned value is a callback that you can executed when you want to update the corresponding metric.

If you need to implement your own custom HTTP endpoints, you can look at our HTTP server, nicknamed "harbor". The docs are here: https://www.liquidsoap.info/doc-2.1.4/harbor_http.html

It is worth noting that the HTTP server's API has been reworked in our current `2.2.x` development cycle. The new API is closer to node's express API and should be easier to use for anyone familiar with it: https://www.liquidsoap.info/doc-dev/harbor_http.html#nodeexpress-api

The `2.2.x` branch is currently being stabilized for release so, if you can tolerante minor hicups that we will try and fix as quickly as possible, it would be advisable to try with this version! You can find release assets at: https://github.com/savonet/liquidsoap/releases/tag/rolling-release-v2.2.x

Lastly, to output values, you can pick the relevant methods on your scripts and use them in custom endpoints and/or prometheus metrics.

See for instance this example that returns the `is_ready` status of a source, implemented with the node-like API:
```liquidsoap
s = ...

def is_s_ready(_, response) =
  response.json({ ready = s.is_ready() })
end

harbor.http.register(port=8000, method="GET", "/is_s_ready", is_s_ready)
```

Hope this help!

**Reply by KlaasT** [reply]
Hi @toots 

thanks so much for the response and sorry for not reacting for two weeks now. Just been way to busy.

I think that is going into the right direction indeed, so thanks for these hints and I hope I can create something with it.

My current setup is as follows:

1. Input which takes an HTTP source for example.
2. 1 - n icecast outputs which distributes the signal to several endpoints.

My problem is that I don't have the endpoint under control so sometimes some might fail for several reasons (password changed, busy, unreachable and so on). That's why I would like to monitor the outputs. Currently I am parsing the logs but log parsers are, in my opinion, just a hacked workaround.

I already thought about splitting up the outputs into several scripts but I fear the encoding overhead when each one has to do their "own" encoding instead of just sharing the source.
